server/index.js

```diff
@@ -68,6 +68,86 @@ app.get('/api/products/:productId', (req, res, next) => {
     .catch(err => next(err));
 });
 
+app.get('/api/cart', (req, res, next) => {
+  const sql = `
+    select *
+    from "carts"`;
+  db.query(sql)
+    .then(result => {
+      const carts = result.rows;
+      res.status(200).json(carts);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/cart', (req, res, next) => {
+  const productId = parseInt(req.body.productId);
+  if (!productId) {
+    res.status(400).json({
+      error: 'Please enter a productId number'
+    });
+    return;
+  }
+  const sqlPrice = `
+  select "price"
+    from "products"
+   where "productId" = $1`;
+  const value = [productId];
+  db.query(sqlPrice, value)
+    .then(priceResult => {
+      if (!priceResult.rows.length) {
+        res.status(400).json({ error: 'No product with this ProductId exists' }
+        );
+      }
+      if (req.session.cartId) {
+        return {
+          cartId: req.session.cartId,
+          price: priceResult.rows[0].price
+        };
+      } else {
+        const sqlNewCart = `
+        insert into "carts" ("cartId", "createdAt")
+        values  (default, default)
+        returning "cartId"`;
+        return db.query(sqlNewCart)
+          .then(cartResult => {
+            return {
+              cartId: cartResult.rows[0].cartId,
+              price: priceResult.rows[0].price
+            };
+          });
+      }
+    })
+    .then(result => {
+      req.session.cartId = result.cartId;
+      const newCartItemRow = `
+      insert into "cartItems" ("cartId", "productId", "price")
+      values ($1, $2, $3)
+      returning "cartItemId"`;
+      const values = [result.cartId, productId, result.price];
+      return db.query(newCartItemRow, values);
+    })
+    // Q: why does huge object return and not just the cartItemId (...returning "cartItemId"`)?
```

Comment:
`db.query()` always returns a promise for a `Result` object. 

https://node-postgres.com/api/result

if you add a `.then()` to your `db.query()`, you'll have the opportunity to pick out just the piece(s) you want, just like you did in the previous `.then()` callback where you put just the `cartId` and `price` in an object.

Reply:
Gotcha. makes sense. thank you!!